extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/CatalogManager.java

```diff
@@ -0,0 +1,202 @@
+
+    /**
+     * An existing catalog table entry was updated.
+     */
+    void updated(TableMetadata table);
+
+    /**
+     * An existing catalog table entry was deleted.
+     */
+    void deleted(TableId id);
+  }
+
+  /**
+   * Start the catalog manager within a Druid run. Called from lifecycle
+   * management and when a coordinator becomes the leader node.
+   */
+  void start();
+
+  /**
+   * Register a listener for catalog events.
+   */
+  void register(Listener listener);
+
+  /**
+   * Create a table entry.
+   *
+   * @return the version of the newly created table. Call
+   * {@link TableMetadata#asUpdate(long)} if you want a new
+   * {@link TableMetadata} with the new version.
+   * @throws {@link DuplicateKeyException} if the row is a duplicate
+   * (schema, name) pair. This generally indicates a code error,
+   * or since our code is perfect, a race condition or a DB
+   * update outside of Druid. In any event, the error is not
+   * retryable: the user should pick another name, or update the
+   * existing table
+   */
+  long create(TableMetadata table) throws DuplicateKeyException;
+
+  /**
+   * Update a table definition.
+   * <p>
+   * If {@code oldVersion == 0}, overwrites any current content.
+   * This is a potential race conditions if this is a partial update
+   * because of the possibility of another user doing an update since the
+   * read. Fine when the goal is to replace the entire definition.
+   * Else, only does the update is at the given version.
+   * <p>
+   * Retryable only if the version is given, and an
+   * {@code OutOfDateException} is thrown.
+   */
+  long update(TableMetadata table, long oldVersion) throws OutOfDateException, NotFoundException;
+
+  /**
+   * Update the table spec incrementally using the transform provided. Performs the update
+   * in a transaction to ensure the read and write are atomic.
+   *
+   * @param id        the table to update
+   * @param transform the transform to apply to the table spec
+   * @return          the update timestamp (version) of the updated record
+   */
+  long updatePayload(TableId id, Function<TableSpec, TableSpec> transform) throws NotFoundException;
+
+  /**
+   * Move the table to the deleting state. No version check: fine
+   * if the table is already in the deleting state. Does nothing if the
+   * table does not exist.
+   *
+   * @return new table update timestamp, or 0 if the table does not
+   * exist
+   */
+  long markDeleting(TableId id);
+
+  /**
+   * Read the table record for the given ID.
+   *
+   * @return the table record, or {@code null} if the entry is not
+   * found in the DB.
+   */
+  @Nullable TableMetadata read(TableId id);
+
+  /**
+   * Delete the table record for the given ID. Essentially does a
+   * "DELETE IF EXISTS". There is no version check. Delete should be
+   * called only when there are no segments left for the table: use
+   * {@link #markDeleting(TableId)} to indicates that the segments are
```

Comment:
To clarify, is the plan for calling the markDeleting method to trigger segment deletion for the underlying datasource?

Reply:
Makes sense. Thanks for clarifying